File: lib/models/event_detail.dart
/// Represents an event schedule item.
class EventSchedule {
  final int id;
  final int eventId;
  final String title;
  final DateTime startTime;
  final DateTime? endTime;
  final String? description;

  EventSchedule({
    required this.id,
    required this.eventId,
    required this.title,
    required this.startTime,
    this.endTime,
    this.description,
  });

  factory EventSchedule.fromJson(Map<String, dynamic> json) {
    return EventSchedule(
      id: json['id'],
      eventId: json['event'],
      title: json['title'],
      startTime: DateTime.parse(json['start_time']),
      endTime: json['end_time'] != null ? DateTime.parse(json['end_time']) : null,
      description: json['description'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'event': eventId,
      'title': title,
      'start_time': startTime.toIso8601String(),
      'end_time': endTime?.toIso8601String(),
      'description': description,
    };
  }

  /// Formatted time range
  String get formattedTime {
    final timeFormat = 'HH:mm';
    if (endTime == null) {
      return startTime.toString().split(' ')[1].substring(0, 5);
    }
    return '${startTime.toString().split(' ')[1].substring(0, 5)} - ${endTime!.toString().split(' ')[1].substring(0, 5)}';
  }
}

/// Represents an aid station.
class AidStation {
  final int id;
  final int eventId;
  final String name;
  final double kilometerMarker;
  final String supplies;
  final bool isMedical;

  AidStation({
    required this.id,
    required this.eventId,
    required this.name,
    required this.kilometerMarker,
    required this.supplies,
    required this.isMedical,
  });

  factory AidStation.fromJson(Map<String, dynamic> json) {
    return AidStation(
      id: json['id'],
      eventId: json['event'],
      name: json['name'],
      kilometerMarker: (json['kilometer_marker'] as num).toDouble(),
      supplies: json['supplies'],
      isMedical: json['is_medical'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'event': eventId,
      'name': name,
      'kilometer_marker': kilometerMarker,
      'supplies': supplies,
      'is_medical': isMedical,
    };
  }
}

/// Represents a route segment.
class RouteSegment {
  final int id;
  final int eventId;
  final int order;
  final String title;
  final String description;
  final double distanceKm;
  final int elevationGain;

  RouteSegment({
    required this.id,
    required this.eventId,
    required this.order,
    required this.title,
    required this.description,
    required this.distanceKm,
    required this.elevationGain,
  });

  factory RouteSegment.fromJson(Map<String, dynamic> json) {
    return RouteSegment(
      id: json['id'],
      eventId: json['event'],
      order: json['order'],
      title: json['title'],
      description: json['description'],
      distanceKm: (json['distance_km'] as num).toDouble(),
      elevationGain: json['elevation_gain'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'event': eventId,
      'order': order,
      'title': title,
      'description': description,
      'distance_km': distanceKm,
      'elevation_gain': elevationGain,
    };
  }
}

/// Represents an event document.
class EventDocument {
  final int id;
  final int eventId;
  final String title;
  final String documentUrl;
  final String documentType;
  final String uploadedBy;
  final DateTime uploadedAt;

  EventDocument({
    required this.id,
    required this.eventId,
    required this.title,
    required this.documentUrl,
    required this.documentType,
    required this.uploadedBy,
    required this.uploadedAt,
  });

  factory EventDocument.fromJson(Map<String, dynamic> json) {
    return EventDocument(
      id: json['id'],
      eventId: json['event'],
      title: json['title'],
      documentUrl: json['document_url'],
      documentType: json['document_type'],
      uploadedBy: json['uploaded_by'],
      uploadedAt: DateTime.parse(json['uploaded_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'event': eventId,
      'title': title,
      'document_url': documentUrl,
      'document_type': documentType,
      'uploaded_by': uploadedBy,
      'uploaded_at': uploadedAt.toIso8601String(),
    };
  }

  /// Get document type display name
  String get documentTypeDisplay {
    switch (documentType) {
      case 'gpx':
        return 'GPX Route';
      case 'guide':
        return 'Race Guide';
      case 'brochure':
        return 'Brochure';
      case 'other':
        return 'Other';
      default:
        return documentType;
    }
  }
}

/// Complete event detail response
class EventDetail {
  final List<EventSchedule> schedules;
  final List<AidStation> aidStations;
  final List<RouteSegment> routeSegments;
  final List<EventDocument> documents;

  EventDetail({
    required this.schedules,
    required this.aidStations,
    required this.routeSegments,
    required this.documents,
  });

  factory EventDetail.fromJson(Map<String, dynamic> json) {
    return EventDetail(
      schedules: (json['schedules'] as List? ?? [])
          .map((item) => EventSchedule.fromJson(item))
          .toList(),
      aidStations: (json['aid_stations'] as List? ?? [])
          .map((item) => AidStation.fromJson(item))
          .toList(),
      routeSegments: (json['route_segments'] as List? ?? [])
          .map((item) => RouteSegment.fromJson(item))
          .toList(),
      documents: (json['documents'] as List? ?? [])
          .map((item) => EventDocument.fromJson(item))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'schedules': schedules.map((item) => item.toJson()).toList(),
      'aid_stations': aidStations.map((item) => item.toJson()).toList(),
      'route_segments': routeSegments.map((item) => item.toJson()).toList(),
      'documents': documents.map((item) => item.toJson()).toList(),
    };
  }
}

File: reference/static/js/event_detail.js
document.addEventListener("DOMContentLoaded", () => {
    console.log("Event detail JS loaded");
    const container = document.querySelector(".event-overview");
    if (!container) {
        console.log("Event overview container not found");
        return;
    }
    console.log("Event overview container found");

    const detailEndpoint = container.dataset.detailEndpoint;
    const availabilityEndpoint = container.dataset.availabilityEndpoint;
    const routeContainer = container.querySelector("[data-route-container]");
    const aidContainer = container.querySelector("[data-aid-container]");
    const scheduleContainer = container.querySelector("[data-schedule-container]");
    const capacityRemaining = container.querySelector("[data-capacity-remaining]");
    const capacityRatio = container.querySelector("[data-capacity-ratio]");
    const availabilityLabel = container.querySelector("[data-availability-label]");
    const progressBar = container.querySelector(".progress-bar");
    const progressFill = container.querySelector(".progress-fill");

    // Modal elements
    const modal = document.getElementById("registration-modal");
    const openModalBtn = document.querySelector("[data-open-registration-modal]");
    const closeModalBtn = document.querySelector("[data-close-modal]");
    const registrationForm = document.getElementById("modal-registration-form");

    const formatDate = (isoString, options = {}) => {
        if (!isoString) {
            return "";
        }
        const date = new Date(isoString);
        return date.toLocaleString(undefined, { ...options });
    };

    const renderSegments = (segments) => {
        if (!routeContainer) {
            return;
        }
        if (!segments.length) {
            routeContainer.innerHTML = "<p>No route segments recorded yet.</p>";
            return;
        }
        routeContainer.innerHTML = segments
            .map(
                (segment) => `
                <article>
                    <div class="segment-order">${segment.order}</div>
                    <div>
                        <h3>${segment.title}</h3>
                        <p class="distance">${segment.distance_km} KM &middot; ${segment.elevation_gain} m elevation</p>
                        <p>${segment.description}</p>
                    </div>
                </article>
            `
            )
            .join("");
    };

    const renderAidStations = (stations) => {
        if (!aidContainer) {
            return;
        }
        if (!stations.length) {
            aidContainer.innerHTML = "<p>No aid stations posted yet.</p>";
            return;
        }
        aidContainer.innerHTML = stations
            .map(
                (station) => `
                <article class="station">
                    <h3>${station.name}</h3>
                    <p class="marker">${station.kilometer_marker} KM</p>
                    <p>${station.supplies}</p>
                    ${station.is_medical ? '<span class="badge accent">Medical</span>' : ""}
                </article>
            `
            )
            .join("");
    };

    const renderSchedule = (items) => {
        if (!scheduleContainer) {
            return;
        }
        if (!items.length) {
            scheduleContainer.innerHTML = "<li>No schedule details available yet.</li>";
            return;
        }
        scheduleContainer.innerHTML = items
            .map(
                (item) => `
                <li>
                    <div class="time">${formatDate(item.start_time, {
                        month: "short",
                        day: "numeric",
                        hour: "2-digit",
                        minute: "2-digit",
                    })}</div>
                    <div>
                        <h3>${item.title}</h3>
                        ${
                            item.end_time
                                ? `<p class="time-range">${formatDate(item.end_time, { hour: "2-digit", minute: "2-digit" })}</p>`
                                : ""
                        }
                        <p>${item.description || ""}</p>
                    </div>
                </li>
            `
            )
            .join("");
    };

    const fetchDetail = () => {
        if (!detailEndpoint) {
            return;
        }
        fetch(detailEndpoint, { headers: { "X-Requested-With": "XMLHttpRequest" } })
            .then((response) => {
                if (!response.ok) {
                    throw new Error(`Detail request failed with status ${response.status}`);
                }
                return response.json();
            })
            .then((data) => {
                renderSegments(data.route_segments || []);
                renderAidStations(data.aid_stations || []);
                renderSchedule(data.schedules || []);
            })
            .catch(() => {
                if (routeContainer) {
                    routeContainer.innerHTML = "<p>Unable to load detailed route information at the moment.</p>";
                }
            });
    };

    const updateAvailability = () => {
        if (!availabilityEndpoint) {
            return;
        }
        fetch(availabilityEndpoint, { headers: { "X-Requested-With": "XMLHttpRequest" } })
            .then((response) => {
                if (!response.ok) {
                    throw new Error(`Availability request failed with status ${response.status}`);
                }
                return response.json();
            })
            .then((data) => {
                if (capacityRemaining) {
                    capacityRemaining.textContent =
                        data.remaining === null ? "Unlimited slots" : `${data.remaining} slots remaining`;
                }
                if (capacityRatio) {
                    capacityRatio.textContent = `${data.capacity_ratio}% capacity`;
                }
                if (progressBar && progressFill) {
                    progressBar.setAttribute("aria-valuenow", data.capacity_ratio);
                    progressFill.style.width = `${data.capacity_ratio}%`;
                }
                if (availabilityLabel) {
                    availabilityLabel.textContent = data.is_registration_open
                        ? "Registration is open. Secure your bib today!"
                        : "Registration is currently closed.";
                }
            })
            .catch(() => {
                if (availabilityLabel) {
                    availabilityLabel.textContent = "Unable to refresh registration status at the moment.";
                }
            });
    };

    // Modal functionality
    const showModal = () => {
        if (modal) {
            modal.style.display = "flex";
            document.body.style.overflow = "hidden";
            console.log("Modal shown");
        }
    };

    const hideModal = () => {
        if (modal) {
            modal.style.display = "none";
            document.body.style.overflow = "";
            console.log("Modal hidden");
        }
    };

    const handleFormSubmit = (event) => {
        event.preventDefault();
        const formData = new FormData(registrationForm);
        const submitBtn = registrationForm.querySelector('button[type="submit"]');
        const originalText = submitBtn.textContent;

        // Disable submit button and show loading
        submitBtn.disabled = true;
        submitBtn.textContent = "Submitting...";

        // Get the event slug from the URL or data attribute
        const eventSlug = container.dataset.eventSlug || window.location.pathname.split('/').filter(Boolean)[1];
        console.log("Event slug:", eventSlug);

        fetch(`/register/events/${eventSlug}/register/ajax/`, {
            method: "POST",
            body: formData,
            headers: {
                "X-CSRFToken": document.querySelector('[name=csrfmiddlewaretoken]').value,
                "X-Requested-With": "XMLHttpRequest"
            }
        })
        .then(response => {
            console.log("Response status:", response.status);
            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }
            return response.json();
        })
        .then(data => {
            console.log("Response data:", data);
            if (data.success) {
                // Success - redirect to registration detail
                window.location.href = data.registration_url;
            } else {
                // Handle errors
                const formElement = document.getElementById("modal-registration-form");
                // Clear previous errors
                formElement.querySelectorAll('.error').forEach(el => el.remove());

                // Show non-field errors
                if (data.non_field_errors && data.non_field_errors.length > 0) {
                    const errorDiv = document.createElement('div');
                    errorDiv.className = 'error';
                    errorDiv.innerHTML = data.non_field_errors.join('<br>');
                    formElement.insertBefore(errorDiv, formElement.firstChild);
                }

                // Show field errors
                for (const [fieldName, errors] of Object.entries(data.errors || {})) {
                    const field = formElement.querySelector(`[name="${fieldName}"]`);
                    if (field) {
                        const errorDiv = document.createElement('div');
                        errorDiv.className = 'error';
                        errorDiv.innerHTML = errors.join('<br>');
                        field.parentNode.insertBefore(errorDiv, field.nextSibling);
                    }
                }
            }
        })
        .catch(error => {
            console.error("Registration submission error:", error);
            alert("An error occurred while submitting your registration. Please try again.");
        })
        .finally(() => {
            // Re-enable submit button
            submitBtn.disabled = false;
            submitBtn.textContent = originalText;
        });
    };

    // Event listeners
    if (openModalBtn) {
        console.log("Open modal button found:", openModalBtn);
        openModalBtn.addEventListener("click", showModal);
    } else {
        console.log("Open modal button not found");
    }

    if (closeModalBtn) {
        closeModalBtn.addEventListener("click", hideModal);
    }

    if (modal) {
        // Close modal when clicking outside
        modal.addEventListener("click", (event) => {
            if (event.target === modal) {
                hideModal();
            }
        });

        // Close modal on Escape key
        document.addEventListener("keydown", (event) => {
            if (event.key === "Escape" && modal.style.display === "flex") {
                hideModal();
            }
        });
    }

    if (registrationForm) {
        registrationForm.addEventListener("submit", handleFormSubmit);
    }

    fetchDetail();
    updateAvailability();

    // Refresh availability every 60 seconds to keep data accurate.
    setInterval(updateAvailability, 60000);
});
